zebra-chain/src/sprout/tree.rs

```diff
@@ -9,44 +9,51 @@
 //! append-only.
 //!
 //! A root of a note commitment tree is associated with each treestate.
+
 #![allow(clippy::unit_arg)]
-#![allow(dead_code)]
 
-use std::{collections::VecDeque, fmt};
+use std::{cell::Cell, fmt};
 
 use byteorder::{BigEndian, ByteOrder};
+use incrementalmerkletree::bridgetree::{self, Frontier};
 use lazy_static::lazy_static;
-#[cfg(any(test, feature = "proptest-impl"))]
-use proptest_derive::Arbitrary;
-use sha2::digest::generic_array::GenericArray;
+use thiserror::Error;
 
 use super::commitment::NoteCommitment;
 
-const MERKLE_DEPTH: usize = 29;
+#[cfg(any(test, feature = "proptest-impl"))]
+use proptest_derive::Arbitrary;
+use sha2::digest::generic_array::GenericArray;
 
-/// MerkleCRH^Sprout Hash Function
+/// Sprout note commitment trees have a max depth of 29.
 ///
-/// Used to hash incremental Merkle tree hash values for Sprout.
+/// <https://zips.z.cash/protocol/protocol.pdf#constants>
+pub(super) const MERKLE_DEPTH: usize = 29;
+
+/// MerkleCRH^Sprout Hash Function.
 ///
-/// MerkleCRH^Sprout(layer, left, right) := SHA256Compress(left || right)
+/// Creates nodes of the note commitment tree.
 ///
-/// `layer` is unused for Sprout but used for the Sapling equivalent.
+/// Definition: https://zips.z.cash/protocol/protocol.pdf#merklecrh.
+/// MerkleCRH^Sprout(layer, left, right) := SHA256Compress(left || right).
 ///
-/// https://zips.z.cash/protocol/protocol.pdf#merklecrh
+/// Note: the implementation of MerkleCRH^Sprout does not use the `layer`
+/// argument from the definition above since the argument does not affect the output.
```

Comment:
```suggestion
/// argument from the definition above since the argument does not affect the output.
///
/// [MerkleCRH^Sprout]: https://zips.z.cash/protocol/protocol.pdf#merklecrh.
```